acmed/config/nginx_hooks.toml

```diff
@@ -0,0 +1,130 @@
+type = ["challenge-http-01", "post-operation"]
+cmd = "cat"
+args = [ "-" ]
+stdin_str = """###
+# Let's Encrypt: acme-challenge location
+###
+
+location ^~ /.well-known/acme-challenge/ {
+  allow all;
+  root "{{#if env.HTTP_ROOT}}{{env.HTTP_ROOT}}{{else}}/var/lib/acmed/domains{{/if}}";
+  default_type "text/plain";
+  try_files $uri =404;
+}
+"""
+stdout = "{{#if env.NGINX_CONFDIR}}{{env.NGINX_CONFDIR}}{{else}}/etc/nginx/conf.d{{/if}}/{{#if env.NGINX_CHALLENGE_LOCATION}}{{env.NGINX_CHALLENGE_LOCATION}}{{else}}001-challenge-letsencrypt.conf{{/if}}"
+
+[[hook]]
+name = "nginx-config-challenge-location-echo"
+type = ["challenge-http-01", "post-operation"]
+cmd = "echo"
+args = [
+    "{{#if env.NGINX_CONFDIR}}{{env.NGINX_CONFDIR}}{{else}}/etc/nginx/conf.d{{/if}}/{{#if env.NGINX_CHALLENGE_LOCATION}}{{env.NGINX_CHALLENGE_LOCATION}}{{else}}001-challenge-letsencrypt.conf{{/if}}"
+]
+allow_failure = true
+
+[[group]]
+name = "nginx-config-challenge-location"
+# hook execution in order of definition
+hooks = [
+    "nginx-config-challenge-location-create",
+    "nginx-config-challenge-location-echo",
+    "nginx-config-challenge-location-chmod"
+]
+
+###
+# nginx reference to TLS certificates
+# env NGINX_TLS_CERTIFICATE -> 002-tls-certificates.conf
+# global: {{certificates_directory}}
+# certificate: {{name}}_{{key_type}}.{{file_type}}.{{ext}}
+###
+
+[[hook]]
+name = "nginx-config-certificate-location-chmod"
+type = ["challenge-http-01", "post-operation"]
+cmd = "chmod"
+args = [
+    "ug+rw",
+    "{{#if env.NGINX_CONFDIR}}{{env.NGINX_CONFDIR}}{{else}}/etc/nginx/conf.d{{/if}}/{{#if env.NGINX_TLS_CERTIFICATE}}{{env.NGINX_TLS_CERTIFICATE}}{{else}}002-tls-certificates.conf{{/if}}"
+]
+allow_failure = true
+
+[[hook]]
+name = "nginx-config-certificate-location-create"
+type = ["challenge-http-01", "post-operation"]
+cmd = "cat"
+args = [ "-" ]
+## TODO
+# make following handlebars accessible in post-operation
+# ssl_certificate {{certificates_directory}}/{{name}}_{{key_type}}.crt.{{ext}};
+# ssl_certificate_key {{certificates_directory}}/{{name}}_{{key_type}}.pk.{{ext}};
+# workaround: define a env:ACMED_CERTS pointing to {{certificates_directory}}
+##
+stdin_str = """###
+# Let's Encrypt TLS certificats
+###
+
+ssl_certificate {{#if env.ACMED_CERTS}}{{env.ACMED_CERTS}}{{else}}/var/lib/acmed/certs{{/if}}/{{identifiers.[0]}}_{{key_type}}.crt.pem; # managed by ACMEd
+ssl_certificate_key {{#if env.ACMED_CERTS}}{{env.ACMED_CERTS}}{{else}}/var/lib/acmed/certs{{/if}}/{{identifiers.[0]}}_{{key_type}}.pk.pem; # managed by ACMEd
+
+ssl_session_cache shared:le_nginx_SSL:10m;
+ssl_session_timeout 1440m;
+ssl_session_tickets off;
+
+ssl_protocols TLSv1.2 TLSv1.3;
+ssl_prefer_server_ciphers off;
+
+ssl_ciphers 'ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-PO
+LY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384';
+"""
```

Comment:
If you want a correct TLS configuration, maybe there should be a few other things, like OCSP Stapling (see `ssl_stapling`, `ssl_stapling_verify` and `ssl_trusted_certificate`). Disabling of both gzip and 0-RTT (`ssl_early_data`) is also highly recommended.

Some other parameters like the HSTS and Expect-CT headers are very important but are quite delicate to add.

Reply:
well, i think this is out of scope for acmed. It's up to the webadmin to configure the behaviour of his `https` instance to comply with `best effort` . There are quite a lot good documents that will gide you to configure it well.
Especilly HSTS is somethink that can't be reverted easily, once it is advertised to the web. What do you think?
I would opt in, to make suggestions in the documentation / wiki.

Reply:
Precisely. That's one of the reasons why I'm not very fond of including anything related to TLS configuration in ACMEd. Either it should entirely handle it or not at all. Having something that works but isn't fully configured will only encourage users not do look further how they can/should configure TLS.

Reply:
kind of. 
I would prefer to divide the puzzle into three parts:

## TLS integration tasks
* configurartion of the webserver with TLS `best effort` syntax/functionality
conformant 
* inclusion of TLS certificates
* preperaton of a .well-known location

## Assumption
* daemons
For the majority of users, `webserver` and `acmed` will run on the same hosts. To adopt acmed for this kind of admins, it is an imense struggle, if everything has to be done handcrafted.
The hook might help, to minimize the needed non-automated task.

Precisely:
include to template names into the webserver config, which can be accieved via simple `copy and  paste`  predefined names from the documentation.
This should work for `apache` as well as `nginx`. 

* website config
Every configured website is tailored to provide the root for a `dedicated application` (wiki, nextcloud, company web, etc.). Admins most definetly have to get in line with the upstream documentation and adopt the location parts, etc.

In 2021 `TLS` save websites should be the default. And all TLS settings will/should be documented upstream. Supporting the admins, we might document some guildlines/references in the `acmed wiki` or reference good external docs (TL;DR way).

Essential for the inclusion of `acmed` managed certificates (at least for me: that is the glance of the client daemon) is the domain centered inclusion of the cert and the pub key. Right?
Lets make this process as clean and simple as possible. All the rest will be adopted from the `interested` admins. Everybody else who is expecting `out of the box` automated setup will stay with certbot, et. al.

Does that makes sense to your?